asserts/snap_resource_asserts.go

```diff
@@ -179,3 +179,141 @@ func assembleSnapResourceRevision(assert assertionBase) (Assertion, error) {
+// global-upload (naming.DefaultProvenance)).
+func (respair *SnapResourcePair) Provenance() string {
+	return respair.HeaderString("provenance")
+}
+
+// DeveloperID returns the id of the developer that submitted the snap resource for the snap revision.
+func (respair *SnapResourcePair) DeveloperID() string {
+	return respair.HeaderString("developer-id")
+}
+
+// Timestamp returns the time when the snap-resource-pair was issued.
+func (respair *SnapResourcePair) Timestamp() time.Time {
+	return respair.timestamp
+}
+
+// Implement further consistency checks.
+func (respair *SnapResourcePair) checkConsistency(db RODatabase, acck *AccountKey) error {
+	otherProvenance := respair.Provenance() != naming.DefaultProvenance
+	if !otherProvenance && !db.IsTrustedAccount(respair.AuthorityID()) {
+		// delegating global-upload revisions is not allowed
+		return fmt.Errorf("snap-resource-pair assertion for snap id %q is not signed by a store: %s", respair.SnapID(), respair.AuthorityID())
+	}
+	_, err := db.Find(AccountType, map[string]string{
+		"account-id": respair.DeveloperID(),
+	})
+	if errors.Is(err, &NotFoundError{}) {
+		return fmt.Errorf("snap-resource-pair assertion for snap id %q does not have a matching account assertion for the developer %q", respair.SnapID(), respair.DeveloperID())
+	}
+	if err != nil {
+		return err
+	}
+	a, err := db.Find(SnapDeclarationType, map[string]string{
+		"series":  release.Series,
+		"snap-id": respair.SnapID(),
+	})
+	if errors.Is(err, &NotFoundError{}) {
+		return fmt.Errorf("snap-resource-pair assertion for snap id %q does not have a matching snap-declaration assertion", respair.SnapID())
+	}
+	if err != nil {
+		return err
+	}
+	if otherProvenance {
+		decl := a.(*SnapDeclaration)
+		ras := decl.RevisionAuthority(respair.Provenance())
+		// check that there's matching delegation using the snap-revision
+		matchingRevAuthority := false
+		for _, ra := range ras {
+			// model==store==nil, we do not perform device-specific
+			// checks at this level, those are performed at
+			// higher-level guarding installing actual components
+			if err := ra.checkProvenanceAndRevision(respair, "snap", respair.SnapRevision(), nil, nil); err == nil {
+				matchingRevAuthority = true
+				break
+			}
+		}
+		if !matchingRevAuthority {
+			return fmt.Errorf("snap-resource-pair assertion with provenance %q for snap id %q is not signed by an authorized authority: %s", respair.Provenance(), respair.SnapID(), respair.AuthorityID())
+		}
+	}
+	return nil
+}
+
+// expected interface is implemented
+var _ consistencyChecker = (*SnapResourcePair)(nil)
+
+// Prerequisites returns references to this snap-resource-pair's prerequisite assertions.
+func (respair *SnapResourcePair) Prerequisites() []*Ref {
+	return []*Ref{
+		{Type: SnapDeclarationType, PrimaryKey: []string{release.Series, respair.SnapID()}},
+	}
+}
+
+func assembleSnapResourcePair(assert assertionBase) (Assertion, error) {
```

Comment:
Shouldn't this also check `snap-id` and `resource-name` with `checkNotEmptyString`? Or maybe a full validity check?

Reply:
as they are part of the primary key this is done automatically by the framework, you should see this in practice in TestInvalid. Generally/historically we tend not too format validate snap-ids coming from the store vs users, anyway if the snap-id is invalid the cross-check will fail because there will be no matching snap-declaration. Adding format validation for resource-name here and for the previous assertion is not a bad idea though.